### DISCUSSION TITLE
[FEATURE REQUEST] config option to provide alternate defaults for state modules

### DISCUSSION THREAD

**Discussion by nicholasmhughes** [saltstack/salt#67684] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
Organizations often find themselves standardizing their state implementations due to internal factors. A good example is the `show_changes` parameter of the [`file.managed` state](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.file.html#salt.states.file.managed). This parameter defaults to `True`, but organizations might want to globally disable this parameter for security or operational reasons.

**Describe the solution you'd like**
Ideally, this could be accomplished with a minion configuration option.

```yaml
state_default_overrides:
  file.managed:
    show_changes: false
  pkg.installed:
    allow_updates: true
```

The state defaults will be overridden by the options provided in the config, but explicitly setting the parameters in the state block will still be respected.

**Describe alternatives you've considered**
Organizations wanting to force the implementation of certain defaults often use Jinja macros to implement this feature. However, this requires the author of the formula to know about that macro and use it. I've also seen some organizations use linters to ensure that certain standards are followed, but this can be fragile depending on the method.

**Additional context**
I created a really rough POC script to show the intended use:

```python
from pprint import pprint


__opts__ = {
    "state_default_overrides": {
        "test_things": {
            "param1": "two",
            "param2": "one",
        }
    }
}


def apply_default_overrides(state, state_args):
    for key, val in __opts__.get("state_default_overrides", {}).get(state, {}).items():
        if key not in state_args:
            state_args[key] = val
    return state_args


def state_system(state, state_args):
    if __opts__.get("state_default_overrides", {}).get(state):
        state_args = apply_default_overrides(state, state_args)
    # fake loader
    ret = globals()[state](**state_args)
    return ret


def test_things(name, param1="one", param2="two"):
    return {
        "result": True,
        "comment": f"one = {param1} | two = {param2}",
    }


if __name__ == "__main__":
    state = "test_things"
    state_args = {
        "name": "foo",
        "param2": "three",
    }
    pprint(state_system(state, state_args))
```

**Comment by nicholasmhughes** [comment]
If this seems like a reasonable solution, I can start working on the PR.

CC: @dwoz @s0undt3ch @twangboy

**Comment by OrangeDog** [comment]
A lot of state functions (and/or the modules they call) already handle defaults from config themselves. Adding a global version on top sounds like a recipe for a confusing mess.

Related feature request: #54141

**Comment by nicholasmhughes** [comment]
Some functions certainly do handle it themselves. The proposed feature would obviate the need for further sprawl within the minion configuration file. A single option would handle all future requests to add configuration file defaults for state modules.

**Comment by OrangeDog** [comment]
They use, for the most part, `config.get`, so it's not only the minion config that is involved.

**Comment by nicholasmhughes** [comment]
So, you're advocating to keep the workflow such that individual module functions are responsible for retrieving their own configurations and not centralizing that operation? Or do you have another way you'd like to see it done?

**Comment by OrangeDog** [comment]
I'm advocating for caution, that you need to account for everything that already does it by itself, and not add additional unexpected behaviour or confusion. Connection parameters for things is probably the most common case.

**Comment by max-arnold** [comment]
This reminds me various policy checking tools. Could this tool be external to Salt?

**Comment by nicholasmhughes** [comment]
Recording thoughts/comments from 2024-03-21 Open Hour:

* General consensus is that this is a good feature as long as we can limit/mitigate user surprise
* Ideal implementation for this use case is in the state system although making the change at a different point, such as kwarg handling, could allow for it to impact execution modules as well
* Tying into the renderer system could allow the updated parameters to be shown in debug output
* Showing warnings or summary output at the end could limit user surprise

View YouTube recording for details and attribution.

**Comment by nicholasmhughes** [comment]
> This reminds me various policy checking tools. Could this tool be external to Salt?

@max-arnold , something like salt-lint can catch things like this sometimes. however, heavy Jinja or macro usage often prevents the regex matching from occurring properly. to me, it's the difference between a development process change that has a technical guardrail vs a technical solution with a troubleshooting process change.
